src/clr-angular/forms/common/form.ts

```diff
@@ -28,4 +52,71 @@ export class ClrForm {
   markAsTouched() {
     this.markControlService.markAsTouched();
   }
+
+  @ContentChildren(ClrLabel, { descendants: true })
+  labels: QueryList<ClrLabel>;
+
+  @HostListener('submit')
+  onFormSubmit() {
+    this.markAsTouched();
+    if (isPlatformBrowser(this.platformId)) {
+      this.invalidControls = Array.from(this.el.nativeElement.querySelectorAll('.ng-invalid'));
+      if (this.invalidControls.length > 0) {
+        this.invalidControls[0].focus();
+        this.updateAriaLive();
+      } else {
+        this.resetAriaLive();
+      }
+    }
+  }
+
+  ngOnInit() {
+    this.createAriaLiveElement();
+  }
+
+  private resetAriaLive() {
+    this.ariaLiveElement.innerHTML = '';
+  }
+
+  private updateAriaLive(): void {
+    if (this.invalidControls.length === 0) {
+      return;
+    }
+
+    this.ariaLiveText = this.commonStrings.keys.formErrorSummary;
+    this.resetAriaLive();
+
+    this.ariaLiveElement.textContent =
+      this.commonStrings.keys.formErrorSummary +
+      this.labels
+        .filter(label => label.forAttr !== undefined)
+        .filter(label => this.invalidControls.find(control => label.forAttr === control.id))
```

Comment:
We don't really need to use filtering twice for the same `label.forAttr`. As far as I see, the first filtering is unnecessary. If it's `undefined`, this check `label.forAttr === control.id` would weed out it just as well.

Reply:
will remove.